python/dgl/sampling/labor.py

```diff
@@ -0,0 +1,276 @@
+                frontier, importances = eid_excluder(frontier, importances)
+            else:
+                frontier = eid_excluder(frontier)
+    return (frontier, importances) if output_device is None else (frontier.to(output_device), list(map(lambda x: x.to(output_device), importances)))
+
+def _sample_labors(g, nodes, fanout, random_seed, edge_dir='in', prob=None,
+                      importance_sampling=0, copy_ndata=True, copy_edata=True, exclude_edges=None):
+    if not isinstance(nodes, dict):
+        if len(g.ntypes) > 1:
+            raise DGLError("Must specify node type when the graph is not homogeneous.")
+        nodes = {g.ntypes[0] : nodes}
+
+    nodes = utils.prepare_tensor_dict(g, nodes, 'nodes')
+    if len(nodes) == 0:
+        raise ValueError(
+            "Got an empty dictionary in the nodes argument. "
+            "Please pass in a dictionary with empty tensors as values instead.")
+    ctx = utils.to_dgl_context(F.context(next(iter(nodes.values()))))
+    nodes_all_types = []
+    NIDs_all_types = []
+    for ntype in g.ntypes:
+        if ntype in nodes:
+            nodes_all_types.append(F.to_dgl_nd(nodes[ntype]))
+            try:
+                if len(g.ntypes) > 1:
+                    NIDs_all_types.append(F.to_dgl_nd(g.ndata[NID][ntype]))
+                else:
+                    NIDs_all_types.append(F.to_dgl_nd(g.ndata[NID]))
+            except:
+                NIDs_all_types.append(nd.array([], ctx=ctx))
+        else:
+            nodes_all_types.append(nd.array([], ctx=ctx))
+            NIDs_all_types.append(nd.array([], ctx=ctx))
+
+    if isinstance(fanout, nd.NDArray):
+        fanout_array = fanout
+    else:
+        if not isinstance(fanout, dict):
+            fanout_array = [int(fanout)] * len(g.etypes)
+        else:
+            if len(fanout) != len(g.etypes):
+                raise DGLError('Fan-out must be specified for each edge type '
+                               'if a dict is provided.')
+            fanout_array = [None] * len(g.etypes)
+            for etype, value in fanout.items():
+                fanout_array[g.get_etype_id(etype)] = value
+        fanout_array = F.to_dgl_nd(F.tensor(fanout_array, dtype=F.int64))
+    
+    if isinstance(prob, list) and len(prob) > 0 and \
+            isinstance(prob[0], nd.NDArray):
+        prob_arrays = prob
+    elif prob is None:
+        prob_arrays = [nd.array([], ctx=nd.cpu())] * len(g.etypes)
+    else:
+        prob_arrays = []
+        for etype in g.canonical_etypes:
+            if prob in g.edges[etype].data:
+                prob_arrays.append(F.to_dgl_nd(g.edges[etype].data[prob]))
+            else:
+                prob_arrays.append(nd.array([], ctx=nd.cpu()))
+
+    excluded_edges_all_t = []
+    if exclude_edges is not None:
+        if not isinstance(exclude_edges, dict):
+            if len(g.etypes) > 1:
+                raise DGLError("Must specify etype type when the graph is not homogeneous.")
```

Comment:
"etype"

Reply:
I copied this code from python/dgl/sampling/neighbor.py, should I change that file too?

Reply:
Yes please.